docs/source/app-dev/app-arch.rst

```diff
@@ -133,22 +133,132 @@ Command deduplication
+The application keeps track of the in-flight command submissions via their submission IDs, but need not persist them.
+
+#. After starting up, the application submits a dummy command with a fresh command ID and submission ID (e.g., a random UUID), but no submission rank or deduplication period.
+   This should generate a definite-answer completion with a completion offset ``off_start``.
+   The application can use the command service for the submission and extract the completion offset from the successful response or the error details of a rejection.   
+   If the submission fails without generating a completion offset, the application should repeat this step until it receives such a fresh completion offset.
+
+#. All subsequent command submission should use a submission rank that is between ``off_start`` and the current completion end.
+   To that end, we assume that the application maintains a running completion offset ``off_running``.
+   It is initialized to ``off_start`` and need not be persisted.
+   Whenever the application observes a completion offset on the completion stream or as part of an RPC response from the command service,
+   it advances ``off_running`` to the observed offset.
+
+#. Starting above the watermark, the application (re)submits the persisted changes and new incoming changes.
+   
+   - The command ID is derived deterministically from the intended change, e.g., using a hash function.
+   - A fresh submission ID is set (e.g., a random UUID).
+     Each resubmission should use a fresh submission ID.
+   - The submission rank is set to ``off_running``.
+   - The deduplication period should start before the first time that the application might have submitted the command.
+     This can be approximated by the watermark's time minus the processing latency, clock skews, and message delays.
+   
+   The following errors around command deduplication should be dealt with (see ``error.proto`` for details on the errors):
+
+   - ``NOT_FOUND`` for the deduplication start:
+     The specified deduplication period is too long.
+     The application should find a shorter deduplication time that is still safe.
+     Alternatively, the application can resubmit the change ID with a dummy command instead of the intended ledger change;
+     when it receives a definite answer for the dummy and there are no other submissions with a higher submission rank,
+     it can conclude that the actual ledger change will not be applied.
+
+   - ``OUT_OF_RANGE`` for the deduplication start:
+     The deduplication start lies in the future.
+     Resubmit with an earlier deduplication start.
+
+   - ``OUT_OF_RANGE`` for the submission rank:
+     This should happen only when the application uses a completion offset obtained from one participant for a submission on another participant.
+     The application should not mix completion offsets from different participants.
```

Comment:
Even in HA setups or are you only referring to different logical participants?

Reply:
We have various HA setups (VMBC's logical participants, Canton's replicated participant) with different failover stories and transparency of the replication. So we probably can't have rules here that work for every ledger (as some ledgers don't even provide stable offsets across participants). I've added a sentence for the failover scenario.